REGRESSION (277389@main): [iOS] AutoFill not offered on autofocus with an attached hardware keyboard

pxlcoder opened the pull request:
#### e1e849f259d0a32e603e644d6fc6b6ef8774ed22
<pre>
REGRESSION (277389@main): [iOS] AutoFill not offered on autofocus with an attached hardware keyboard
<a href="https://bugs.webkit.org/show_bug.cgi?id=273086">https://bugs.webkit.org/show_bug.cgi?id=273086</a>
<a href="https://rdar.apple.com/126653124">rdar://126653124</a>

Reviewed by Chris Dumez, Abrar Rahman Protyasha and Richard Robinson.

Focusing a field without user interaction should not offer AutoFill unless there
is a hardware keyboard attached. In order to implement this behavior, Safari
checks the returned value of `GSEventIsHardwareKeyboardAttached()` in their
injected bundle.

Since that call is made in the web process, where notifyd is blocked following
277389@main, it always returns false, as `GSEventIsHardwareKeyboardAttached()`
relies on libnotify. Consequently, AutoFill is not offered until the user
actually taps on a field, even when a hardware keyboard is attached.

Simply updating `WebProcessPool::registerNotificationObservers` to register
for the hardware keyboard notification is not sufficient, as the returned value
of `GSEventIsHardwareKeyboardAttached()` will be incorrect if the process is
launched with the hardware keyboard already attached.

Fix by explicitly setting the state at process creation time, and whenever the
value changes, using `GSEventSetHardwareKeyboardAttachedWithCountryCodeAndType()`.
A new `HardwareKeyboardState` struct is introduced to represent the state across
process boundaries. Additionally suppress updates in the UI process if the state
is unchanged, to avoid notification loops in the simulator.

* Source/WebCore/PAL/pal/spi/ios/GraphicsServicesSPI.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/Shared/ios/HardwareKeyboardState.cpp: Copied from Source/WebCore/PAL/pal/spi/ios/GraphicsServicesSPI.h.
(WebKit::currentHardwareKeyboardState):
(WebKit::setHardwareKeyboardState):
* Source/WebKit/Shared/ios/HardwareKeyboardState.h: Copied from Source/WebCore/PAL/pal/spi/ios/GraphicsServicesSPI.h.
* Source/WebKit/Shared/ios/HardwareKeyboardState.serialization.in: Added.
* Source/WebKit/Shared/mac/MediaFormatReader/MediaSampleCursor.h:

Unified sources build fix.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(hardwareKeyboardAvailabilityChangedCallback):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::hardwareKeyboardAvailabilityChanged):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::platformInitialize):
(WebKit::WebPageProxy::applicationWillEnterForeground):
(WebKit::WebPageProxy::hardwareKeyboardAvailabilityChanged):
(WebKit::WebPageProxy::isInHardwareKeyboardMode): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::platformInitialize):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::hardwareKeyboardAvailabilityChanged):
* WebKitLibraries/SDKs/appletvos16.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/appletvos17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/appletvsimulator16.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/appletvsimulator17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/iphoneos16.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/iphonesimulator16.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/watchos10.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/watchos9.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/watchsimulator10.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/watchsimulator9.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:

Canonical link: <a href="https://commits.webkit.org/277961@main">https://commits.webkit.org/277961@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d926157209bef3d504ee9086b7cc82811ab263d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28324 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/52075 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51830 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/45162 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51389 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34295 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25858 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40108 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/50566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25926 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/52075 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21221 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23381 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/52075 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7283 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/52075 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53744 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/24161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20369 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47430 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/52075 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10794 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/26229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/ios/HardwareKeyboardState.cpp:
```diff
@@ -0,0 +1,51 @@
+    state.isAttached = GSEventIsHardwareKeyboardAttached();
+    state.countryCode = GSEventGetHardwareKeyboardCountry();
+    state.keyboardType = GSEventGetHardwareKeyboardType();
+    return state;
```
Optional suggestion for brevity: If you want to keep the fields explicit, maybe this?
```suggestion
    return {
        .isAttached = GSEventIsHardwareKeyboardAttached(),
        .countryCode = GSEventGetHardwareKeyboardCountry(),
        .keyboardType = GSEventGetHardwareKeyboardType(),
    };
```

Or just drop the designated initializers and go with `return { ..., ..., ... }`

pxlcoder replied:
Making either change ends up introducing this error:

```
error: non-constant-expression cannot be narrowed from type 'Boolean' (aka 'unsigned char') to 'bool' in initializer list [-Wc++11-narrowing]
```

Keeping as-is, unless the explicit cast is also preferred?

aprotyas reviewed (approved):
Makes sense to me. Is there a reason we're keeping track of `countryCode` and `keyboardType`?

aprotyas reviewed (commented):
Editorial comment: Small grammatical error in the change log?
> will be incorrect if the process is launched with the hardware keyboard is already attached.

cdumez commented on Source/WebKit/Shared/ios/HardwareKeyboardState.cpp:
```diff
@@ -0,0 +1,51 @@
+
+HardwareKeyboardState currentHardwareKeyboardState()
+{
+    HardwareKeyboardState state;
```
Would be nice to write this more concisely:
```
return {
    !!GSEventIsHardwareKeyboardAttached(),
    GSEventGetHardwareKeyboardCountry(),
    GSEventGetHardwareKeyboardType()
};
```

cdumez commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1145,10 +1142,18 @@ static FloatSize fullscreenPreferencesScreenSize(CGFloat preferredWidth)
+
     updateCurrentModifierState();
-    m_process->send(Messages::WebPage::HardwareKeyboardAvailabilityChanged(keyboardIsAttached), webPageID());
+    m_process->send(Messages::WebPage::HardwareKeyboardAvailabilityChanged(currentHardwareKeyboardState()), webPageID());
```
Why can't we reuse the `hardwareKeyboardState` local here?

pxlcoder replied:
Will make both changes, as soon as the active `api-ios` EWS run is finished.

cdumez reviewed (approved):
LGTM

pxlcoder committed:
REGRESSION (277389@main): [iOS] AutoFill not offered on autofocus with an attached hardware keyboard
https://bugs.webkit.org/show_bug.cgi?id=273086
rdar://126653124

Reviewed by Chris Dumez, Abrar Rahman Protyasha and Richard Robinson.

Focusing a field without user interaction should not offer AutoFill unless there
is a hardware keyboard attached. In order to implement this behavior, Safari
checks the returned value of `GSEventIsHardwareKeyboardAttached()` in their
injected bundle.

Since that call is made in the web process, where notifyd is blocked following
277389@main, it always returns false, as `GSEventIsHardwareKeyboardAttached()`
relies on libnotify. Consequently, AutoFill is not offered until the user
actually taps on a field, even when a hardware keyboard is attached.

Simply updating `WebProcessPool::registerNotificationObservers` to register
for the hardware keyboard notification is not sufficient, as the returned value
of `GSEventIsHardwareKeyboardAttached()` will be incorrect if the process is
launched with the hardware keyboard already attached.

Fix by explicitly setting the state at process creation time, and whenever the
value changes, using `GSEventSetHardwareKeyboardAttachedWithCountryCodeAndType()`.
A new `HardwareKeyboardState` struct is introduced to represent the state across
process boundaries. Additionally suppress updates in the UI process if the state
is unchanged, to avoid notification loops in the simulator.

* Source/WebCore/PAL/pal/spi/ios/GraphicsServicesSPI.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/Shared/ios/HardwareKeyboardState.cpp: Copied from Source/WebCore/PAL/pal/spi/ios/GraphicsServicesSPI.h.
(WebKit::currentHardwareKeyboardState):
(WebKit::setHardwareKeyboardState):
* Source/WebKit/Shared/ios/HardwareKeyboardState.h: Copied from Source/WebCore/PAL/pal/spi/ios/GraphicsServicesSPI.h.
* Source/WebKit/Shared/ios/HardwareKeyboardState.serialization.in: Added.
* Source/WebKit/Shared/mac/MediaFormatReader/MediaSampleCursor.h:

Unified sources build fix.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(hardwareKeyboardAvailabilityChangedCallback):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::hardwareKeyboardAvailabilityChanged):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::platformInitialize):
(WebKit::WebPageProxy::applicationWillEnterForeground):
(WebKit::WebPageProxy::hardwareKeyboardAvailabilityChanged):
(WebKit::WebPageProxy::isInHardwareKeyboardMode): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::platformInitialize):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::hardwareKeyboardAvailabilityChanged):
* WebKitLibraries/SDKs/appletvos16.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/appletvos17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/appletvsimulator16.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/appletvsimulator17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/iphoneos16.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/iphonesimulator16.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/watchos10.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/watchos9.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/watchsimulator10.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:
* WebKitLibraries/SDKs/watchsimulator9.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd:

Canonical link: https://commits.webkit.org/277961@main
